common/lib/xmodule/xmodule/modulestore/perf_tests/generate_report.py

```diff
@@ -70,105 +73,214 @@ def tostring(self, pretty_print=False):
+
+        self.all_modulestore_combos = set()
+        for row in self.all_rows:
+            time_taken = row[3]
+
+            # Split apart the description into its parts.
+            desc_parts = row[2].split(':')
+            modulestores = desc_parts[1]
+            self.all_modulestore_combos.add(modulestores)
+            amount_md = desc_parts[2]
+            test_phase = 'all'
+            if len(desc_parts) > 3:
+                test_phase = desc_parts[3]
+
+            # Save the data in a multi-level dict - { phase1: { amount1: {ms1->ms2: duration, ...}, ...}, ...}.
+            phase_data = self.run_data.setdefault(test_phase, {})
+            amount_data = phase_data.setdefault(amount_md, {})
+            __ = amount_data.setdefault(modulestores, time_taken)
+
+    def generate_html(self):
+        """
+        Generate HTML.
+        """
+        html = HTMLDocument("Results")
+
+        # Output comparison of each phase to a different table.
+        for phase in self.run_data.keys():
+            if phase in ('fake_assets',):
+                continue
+            per_phase = self.run_data[phase]
+            html.add_header(1, phase)
+
+            title_map = {
+                'duration': 'Total Duration (ms)',
+                'ratio': 'Total Duration Per Number of Assets (ms/asset)',
+                'variable_cost': 'Asset Export Duration Per Number of Assets (ms/asset)'
+            }
+            for table_type in ('duration', 'ratio', 'variable_cost'):
+                if phase == 'all' and table_type in ('ratio', 'variable_cost'):
+                    continue
+                # Make the table header columns and the table.
+                columns = ["Asset Metadata Amount", ]
+                ms_keys = sorted(self.all_modulestore_combos)
+                for k in ms_keys:
+                    columns.append("{} ({})".format(k, table_type))
+                phase_table = HTMLTable(columns)
+
+                # Make a row for each amount of asset metadata.
+                for amount in sorted(per_phase.keys()):
+                    per_amount = per_phase[amount]
+                    num_assets = int(amount)
+                    row = [amount, ]
+                    for modulestore in ms_keys:
+                        if table_type == 'duration':
+                            value = per_amount[modulestore]
+                        elif table_type == 'ratio':
+                            if num_assets != 0:
+                                value = per_amount[modulestore] / float(amount)
+                            else:
+                                value = 0
+                        elif table_type == 'variable_cost':
+                            if num_assets == 0:
+                                value = 0
+                            else:
+                                value = (per_amount[modulestore] - per_phase['0'][modulestore]) / float(amount)
+                        row.append("{}".format(value))
```

Comment:
`value` is a float, which was causing HTML problems. I chose `"".format()` over `str()`.